File: tide_graph.go
// tide_graph.go
package main

import (
	"fmt"
	"math"
)

const (
	amplitude  = 1.5
	meanLevel  = 0.5
	period     = 12.42 // hours
)

func tideLevel(hours float64) float64 {
	rad := (hours / period) * 2 * math.Pi
	return meanLevel + amplitude*math.Sin(rad-1.2)
}

func generateData() [][2]float64 {
	data := make([][2]float64, 0, 24*6)
	for minutes := 0; minutes < 24*60; minutes += 10 {
		h := float64(minutes) / 60.0
		lvl := tideLevel(h)
		data = append(data, [2]float64{h, lvl})
	}
	return data
}

func findExtrema(data [][2]float64) ([][2]float64, [][2]float64) {
	var highs, lows [][2]float64
	for i := 1; i < len(data)-1; i++ {
		prev, curr, next := data[i-1], data[i], data[i+1]
		if curr[1] > prev[1] && curr[1] > next[1] {
			highs = append(highs, curr)
		} else if curr[1] < prev[1] && curr[1] < next[1] {
			lows = append(lows, curr)
		}
	}
	return highs, lows
}

func drawGraph(data [][2]float64) string {
	levels := make([]float64, len(data))
	minLvl, maxLvl := 0.0, 0.0
	for i, d := range data {
		levels[i] = d[1]
		if i == 0 || d[1] < minLvl {
			minLvl = d[1]
		}
		if i == 0 || d[1] > maxLvl {
			maxLvl = d[1]
		}
	}
	// Add padding
	minLvl = math.Floor(minLvl-0.2)
	maxLvl = math.Ceil(maxLvl+0.2)

	height := 20
	scale := (maxLvl - minLvl) / float64(height)

	cols := 50
	step := len(data) / cols
	graphData := make([]struct{ h, lvl float64; row int }, cols)
	for i := 0; i < cols; i++ {
		idx := i * step
		h, lvl := data[idx][0], data[idx][1]
		row := int(math.Round((maxLvl - lvl) / scale))
		if row < 0 {
			row = 0
		}
		if row >= height {
			row = height - 1
		}
		graphData[i] = struct{ h, lvl float64; row int }{h, lvl, row}
	}

	var lines []string
	lines = append(lines, "🌊 Tide Graph (24‑hour forecast)")
	lines = append(lines, "")
	lines = append(lines, "Level (m)")

	for r := 0; r < height; r++ {
		lvl := maxLvl - float64(r)*scale
		line := fmt.Sprintf("%5.1f |", lvl)
		for c := 0; c < cols; c++ {
			if graphData[c].row == r {
				line += "*"
			} else {
				line += " "
			}
		}
		lines = append(lines, line)
	}

	// X-axis
	lines = append(lines, "      +"+strings.Repeat("-", cols))

	// Time labels
	timeLine := "       "
	for i := 0; i <= 24; i += 3 {
		pos := int(float64(i) / 24.0 * float64(cols))
		label := fmt.Sprintf("%02d:00", i)
		for len(timeLine) < pos {
			timeLine += " "
		}
		timeLine += label
	}
	lines = append(lines, timeLine)
	lines = append(lines, "")

	highs, lows := findExtrema(data)
	if len(highs) > 0 {
		str := "High tides: "
		for i, h := range highs[:2] {
			if i > 0 {
				str += ", "
			}
			hour := int(h[0])
			min := int((h[0] - float64(hour)) * 60)
			str += fmt.Sprintf("%02d:%02d (%.1fm)", hour, min, h[1])
		}
		lines = append(lines, str)
	}
	if len(lows) > 0 {
		str := "Low tides:  "
		for i, l := range lows[:2] {
			if i > 0 {
				str += ", "
			}
			hour := int(l[0])
			min := int((l[0] - float64(hour)) * 60)
			str += fmt.Sprintf("%02d:%02d (%.1fm)", hour, min, l[1])
		}
		lines = append(lines, str)
	}

	result := ""
	for _, line := range lines {
		result += line + "\n"
	}
	return result
}

func main() {
	data := generateData()
	graph := drawGraph(data)
	fmt.Print(graph)
}

File: tide_graph.js
// tide_graph.js
const AMPLITUDE = 1.5;
const MEAN_LEVEL = 0.5;
const PERIOD = 12.42;

function tideLevel(hours) {
    const rad = (hours / PERIOD) * 2 * Math.PI;
    return MEAN_LEVEL + AMPLITUDE * Math.sin(rad - 1.2);
}

function generateData() {
    const data = [];
    for (let minutes = 0; minutes < 24*60; minutes += 10) {
        const h = minutes / 60;
        const lvl = tideLevel(h);
        data.push([h, lvl]);
    }
    return data;
}

function findExtrema(data) {
    const highs = [], lows = [];
    for (let i = 1; i < data.length - 1; i++) {
        const prev = data[i-1], curr = data[i], next = data[i+1];
        if (curr[1] > prev[1] && curr[1] > next[1]) highs.push(curr);
        else if (curr[1] < prev[1] && curr[1] < next[1]) lows.push(curr);
    }
    return { highs, lows };
}

function drawGraph(data) {
    const levels = data.map(d => d[1]);
    let minLvl = Math.min(...levels);
    let maxLvl = Math.max(...levels);
    minLvl = Math.floor(minLvl - 0.2);
    maxLvl = Math.ceil(maxLvl + 0.2);

    const height = 20;
    const scale = (maxLvl - minLvl) / height;

    const cols = 50;
    const step = Math.floor(data.length / cols);
    const graphData = [];
    for (let i = 0; i < cols; i++) {
        const idx = i * step;
        const h = data[idx][0];
        const lvl = data[idx][1];
        let row = Math.round((maxLvl - lvl) / scale);
        if (row < 0) row = 0;
        if (row >= height) row = height - 1;
        graphData.push({ h, lvl, row });
    }

    let lines = [];
    lines.push("🌊 Tide Graph (24‑hour forecast)");
    lines.push("");
    lines.push("Level (m)");

    for (let r = 0; r < height; r++) {
        const lvl = maxLvl - r * scale;
        let line = lvl.toFixed(1).padStart(5) + " |";
        for (let c = 0; c < cols; c++) {
            line += graphData[c].row === r ? "*" : " ";
        }
        lines.push(line);
    }

    lines.push("      +" + "-".repeat(cols));

    // Time labels
    let timeLine = "       ";
    for (let i = 0; i <= 24; i += 3) {
        const pos = Math.floor((i / 24) * cols);
        const label = String(i).padStart(2, '0') + ":00";
        while (timeLine.length < pos) timeLine += " ";
        timeLine += label;
    }
    lines.push(timeLine);
    lines.push("");

    const { highs, lows } = findExtrema(data);
    if (highs.length > 0) {
        let str = "High tides: ";
        for (let i = 0; i < Math.min(2, highs.length); i++) {
            if (i > 0) str += ", ";
            const h = highs[i][0];
            const hour = Math.floor(h);
            const min = Math.round((h - hour) * 60);
            str += `${String(hour).padStart(2,'0')}:${String(min).padStart(2,'0')} (${highs[i][1].toFixed(1)}m)`;
        }
        lines.push(str);
    }
    if (lows.length > 0) {
        let str = "Low tides:  ";
        for (let i = 0; i < Math.min(2, lows.length); i++) {
            if (i > 0) str += ", ";
            const h = lows[i][0];
            const hour = Math.floor(h);
            const min = Math.round((h - hour) * 60);
            str += `${String(hour).padStart(2,'0')}:${String(min).padStart(2,'0')} (${lows[i][1].toFixed(1)}m)`;
        }
        lines.push(str);
    }

    return lines.join("\n");
}

console.log(drawGraph(generateData()));
